plugins/modem-manager/fu-mm-device.c

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2018 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include "fu-mm-device.h"
+
+struct _FuMmDevice {
+	FuDevice		 parent_instance;
+	MMObject		*modem;
+};
+
+G_DEFINE_TYPE (FuMmDevice, fu_mm_device, FU_TYPE_DEVICE)
+
+static void
+fu_mm_device_to_string (FuDevice *device, GString *str)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+	g_string_append (str, "  FuMmDevice:\n");
+	g_string_append_printf (str, "    path:\t\t\t%s\n",
+				mm_object_get_path (self->modem));
+}
+
+static gboolean
+fu_mm_device_setup (FuDevice *device, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+	MMModem *mm = mm_object_peek_modem (self->modem);
+	fu_device_set_vendor (device, mm_modem_get_manufacturer (mm));
+	fu_device_set_name (device, mm_modem_get_model (mm));
+	fu_device_set_version (device, mm_modem_get_revision (mm));
+//	fu_device_set_version_bootloader (device, mm_modem_get_hardware_revision (mm));	//FIXME only available in some very new ver
+	fu_device_add_guid (device, mm_modem_get_device_identifier (mm));
+	return TRUE;
+}
+
+static gboolean
+fu_mm_device_detach (FuDevice *device, GError **error)
+{
+	FuMmDevice *self = FU_MM_DEVICE (device);
+	//FIXME: need to call new PrepareForFirmwareUpdate() method on interface
```

Comment:
And probably check whether that method exists at startup (for older MM scenario).  If it doesn't exist set the device as not updateable and set `UpdateError`.

Reply:
libmm-glib will use dbus activation for the system modem manager?

Can you inquire what happens if they get out of sync?

For flatpak and snap would be ideal to build against the correct libmm-glib and then rely on the modem manager outside the sandbox to communicate with.

Reply:
> libmm-glib will use dbus activation for the system modem manager

Yes.

> what happens if they get out of sync?

That's a good question. I assume what happens is the MM daemon gets started up (although, it's going to be started by NM when the session starts before that), we call the method, and we get back `G_DBUS_ERROR_UNKNOWN_METHOD`.

There doesn't appear to be a way to get the runtime version. I've filed https://gitlab.freedesktop.org/mobile-broadband/ModemManager/issues/94 for comments.

Reply:
Sounds good.  Thanks for getting that part filed.  Even with adding a property you'll be able to set `UpdateError` if that property is missing (IE older version).

Reply:
I've added some code in fwupd to test the runtime version; we'll also need to raise the dep to 1.9.0, which means all the CI has to be disabled for this module. Maybe flatpak can be switched to build from master or something.

For the new method, I've opened https://gitlab.freedesktop.org/mobile-broadband/ModemManager/issues/95 so we can track that in MM.

Reply:
1.9.0 is the "dev" version from git master (so there will be 1.9.0 versions built before this commit gets integrated that don't have the new property). This new API will be released officially with MM 1.10.0, although if you need a quick version bump to 1.9.1 just to flag the existence of this new API, we can do that. Also, note that I'd like to work on this integration in the next weeks, and I'm suggesting a MM 1.10 release by end of this month.

Reply:
My gut would be too turn it on for ci in just one build like snap or flatpak for now for that reason until packages are widely available.

Reply:
@aleksander0m I was getting the packaging for snap sorted and had some (related) observations I wanted to check on.
At least with `modemmanager` master I was seeing that `libqmi` and `libmbim` both need releases to actually build against it.

So my questions:
1. Will 1.10.0 MM release also mean there will be new `libqmi` and `libmbim` releases?
2. For the snap it needs to contain a new enough `libmm-glib` library but it won't actually carry modem manager daemon or udev rules, systemd unit or anything else.  Considering this does the `libmm-glib` being placed in the snap need to build against `libqmi`/`libmbim` or just the daemon does that (which will live outside the snap).
3. (Same above question will apply to flatpak)

Reply:
Hey @superm1 
1. Yes. Just before MM 1.10 is released I'll release new libqmi and libmbim stable versions as well, and will bump the minimum required versions of those libs in MM.
2. and 3. The daemon is the one that requires libqmi and libmbim. So you could rebuild MM without libqmi or libmbim and use the generated libmm-glib to talk to a daemon that does use libqmi and libmbim (if that's the question)

Reply:
Perfect, thanks for the confirmation.